### PR TITLE
Use Python 3.11 by default and venv instead of virtualenv

### PR DESCRIPTION
**Pull request by bgruening**
Closes usegalaxy-eu/issues#465.

### CODE REVIEW

File: group_vars/sn07.yml [hunk 1]
```diff
@@ -350,7 +350,7 @@ galaxy_venv_dir: "{{ galaxy_root }}/venv"
 galaxy_job_working_directory: "{{ galaxy_config['galaxy']['job_working_directory'] }}"
 ucsc_build_sites:
 
-galaxy_virtualenv_command: "{{ conda_prefix }}/envs/_galaxy_/bin/virtualenv"
+galaxy_virtualenv_command: "python -m venv"
```
Comment (kysrpex) [inline group_vars/sn07.yml:353]: Given that venv links to system Python

```shell
[jose@BILP51 Downloads]$ python -m venv test_venv
[jose@BILP51 Downloads]$ ls -lah test_venv/bin
insgesamt 48K
drwxr-xr-x 1 jose jose  164 29. Aug 10:59 .
drwxr-xr-x 1 jose jose   56 29. Aug 10:59 ..
-rw-r--r-- 1 jose jose 2,0K 29. Aug 10:59 activate
-rw-r--r-- 1 jose jose  933 29. Aug 10:59 activate.csh
-rw-r--r-- 1 jose jose 2,2K 29. Aug 10:59 activate.fish
-rw-r--r-- 1 jose jose 8,9K 29. Aug 10:59 Activate.ps1
-rwxr-xr-x 1 jose jose  246 29. Aug 10:59 pip
-rwxr-xr-x 1 jose jose  246 29. Aug 10:59 pip3
-rwxr-xr-x 1 jose jose  246 29. Aug 10:59 pip3.11
lrwxrwxrwx 1 jose jose   15 29. Aug 10:59 python -> /usr/bin/python
lrwxrwxrwx 1 jose jose    6 29. Aug 10:59 python3 -> python
lrwxrwxrwx 1 jose jose    6 29. Aug 10:59 python3.11 -> python
```

I'd feel safer using 

`galaxy_virtualenv_command: "{{ conda_prefix }}/envs/_galaxy_/bin/python -m venv"`

File: group_vars/sn07.yml [hunk 2]
```diff
@@ -279,7 +279,7 @@ condor_extra: |
 gie_proxy_dir: "{{ galaxy_root }}/gie-proxy/proxy"
 gie_proxy_git_version: main
 gie_proxy_setup_nodejs: nodeenv
-gie_proxy_virtualenv_command: /usr/bin/python3 -m virtualenv #"{{ pip_virtualenv_command }}"
+gie_proxy_virtualenv_command: /usr/bin/python3 -m venv #"{{ pip_virtualenv_command }}"
```
Comment (kysrpex) [inline group_vars/sn07.yml:282]: For gie then the only change is switching from virtualenv to venv but it still uses system's Python (3.8) right?

File: group_vars/sn07.yml (hunk 1 above)
Comment (bgruening) [reply group_vars/sn07.yml:353]: Who is updating `{{ conda_prefix }}/envs/_galaxy_/` is that 3.11 then?

Comment (kysrpex) [reply group_vars/sn07.yml:353]: @sanjaysrikakulam You're more familiar with the miniconda role.

**Comment by bgruening** [comment]
Please take this one over, I will be on a bike soon.

File: group_vars/sn07.yml (hunk 1 above)
Comment (kysrpex) [reply group_vars/sn07.yml]: Definitely yes.

```text
[jose@BILP51 Downloads]$ conda create -n python_36 python=3.6
[...]
[jose@BILP51 Downloads]$ /home/jose/.mambaforge/envs/python_36/bin/python -m venv venv_python_36
[jose@BILP51 Downloads]$ venv_python_36/bin/python --version
Python 3.6.15

```

File: group_vars/sn07.yml (hunk 2 above)
Comment (sanjaysrikakulam) [reply group_vars/sn07.yml:282]: Yes, always use the absolute path, please.

**Comment by sanjaysrikakulam** [comment]
Also, all the changes are to `sn07` files and not `sn06`.

**Comment by kysrpex** [comment]
> Also, all the changes are to `sn07` files and not `sn06`.

Do you mean that the same changes should be applied to sn06's playbook and group vars or that this changeset is correct?

**Comment by sanjaysrikakulam** [comment]
> > Also, all the changes are to `sn07` files and not `sn06`.
> 
> Do you mean that the same changes should be applied to sn06's playbook and group vars or that this changeset is correct?

We do not have a sn07 host at the moment. Only sn06. Also, the changes @bgruening made and suggested were concerning to sn06. Though the changes made to sn07 now are useful (keep things synced), they will not change or implement @bgruening's plan for sn06. Maybe that will come in a separate PR?

**Comment by kysrpex** [comment]
> > > Also, all the changes are to `sn07` files and not `sn06`.
> > 
> > 
> > Do you mean that the same changes should be applied to sn06's playbook and group vars or that this changeset is correct?
> 
> We do not have a sn07 host at the moment. Only sn06. Also, the changes @bgruening made and suggested were concerning to sn06. Though the changes made to sn07 now are useful (keep things synced), they will not change or implement @bgruening's plan for sn06. Maybe that will come in a separate PR?

Sorry I was thinking somehow of the maintenance node as sn07 (obviously wrong) :sweat_smile: Now it is clear.

**Comment by bgruening** [comment]
I could not find those variables in sn06. Sorry.

File: group_vars/sn07.yml (hunk 2 above)
Comment (kysrpex) [reply group_vars/sn07.yml]: Most likely yes, the virtual environment is actually just [used to control Node.js environment](https://github.com/usegalaxy-eu/ansible-gie-proxy/blob/ca6b025bc06ae8c0de98c339b1c39e1709e86304/tasks/nodejs.yml#L8C1-L27) using [nodeenv](https://pypi.org/project/nodeenv). The proxy itself is written in JavaScript.

**Comment by sanjaysrikakulam** [comment]
I have added the missing variable now.

_**Notes**_
`tiaas_virtualenv_python` variable in sn06 and sn07 group_vars files is not required in latest version of the role (See [here](https://github.com/galaxyproject/ansible-tiaas2/commit/5d0aa528eb27d0cc67bc10f7c1d1db61319704af)). So we can remove this variable when we migrate to the latest version of the TIaaS2 ansible role.

**@bgruening's migration plan:**

1. turn down handlers
2. mv /opt/galaxy/venv /opt/galaxy/venv39
3. recreate a new venv like above
4. run the playbook
5. sync to NFS for celery and others
6. turn down the manually started gunicorn

**_Updated plan:_**
1. turn down handlers
2. mv /opt/galaxy/venv /opt/galaxy/venv39
3. _Merge this PR and run the playbook_
4. sync to NFS for celery and others
5. turn down the manually started gunicorn and remove the venv311 and the gunicorn socket along with the systemd file (if created)

The changes will now use the miniconda role and create a `miniconda` directory in `/opt/` where a `_galaxy_` conda env with python 3.11 will be created. Python from this conda env will then create `venv` for Galaxy. Galaxy's `venv` will be created in `/opt/galaxy/venv`.

**Review by sanjaysrikakulam — APPROVED** [review]
It should work. Please merge this [banner update PR](https://github.com/galaxyproject/galaxy-hub/pull/2149) and then this PR. Make sure to change the name of the existing `venv` directory to `venv38` before running and deploying this playbook.

**Comment by bgruening** [comment]
I had workshops to do this morning and many questions, so will do it tomorrow morning, your midnight.

**Comment by sanjaysrikakulam** [comment]
@bgruening, `tensorflow==2.9.3` is not available for Python 3.11. The [conditional dependency](https://github.com/usegalaxy-eu/galaxy/blob/release_23.1_europe/lib/galaxy/dependencies/conditional-requirements.txt#L51) should be updated to `tensorflow==2.12.0`. However, I am unsure whether the `tool recommendation` tool would work with `tensorflow==2.12.0` version. This is the 1st version (latest version is `2.13.0`) of Tensorflow that supports Python 3.11.

**Comment by bgruening** [comment]
Ok I will keep an eye on it. I guess we would deactivate it if it's not working and ask Anyp to look at it.

**Comment by bgruening** [comment]
And let's go!

**Pull request merged**